File: u2bun/src/runtime/device.ts
import { selectTargetDevice, forwardPort } from "./adb";
import { U2Client } from "./u2client";
import { DeviceOfflineError } from "../errors";

export class DeviceSession {
  public serial: string = "";
  public timeout: number;
  public adbPath?: string;
  public localPort: number;

  private client?: U2Client;

  constructor(serial?: string, timeout: number = 30, adbPath?: string, localPort: number = 9008) {
    if (serial) this.serial = serial;
    this.timeout = timeout;
    this.adbPath = adbPath;
    this.localPort = localPort;
  }

  async connect(): Promise<U2Client> {
    const { target } = await selectTargetDevice(this.serial, this.adbPath);
    this.serial = target.serial;

    try {
      // Forward ADB port 9008
      await forwardPort(this.serial, this.localPort, 9008, this.adbPath);
      const client = new U2Client(this.localPort, this.timeout);
      
      // Ping check
      await client.ping();
      this.client = client;
      return client;
    } catch (firstErr: any) {
      // Retry once after auto-forward
      try {
        await forwardPort(this.serial, this.localPort, 9008, this.adbPath);
        const client = new U2Client(this.localPort, this.timeout);
        await client.ping();
        this.client = client;
        return client;
      } catch (secondErr: any) {
        throw new DeviceOfflineError(
          `Failed to connect uiautomator2 to device '${this.serial}' after retry: ${secondErr.message || String(secondErr)}`
        );
      }
    }
  }

  async getClient(): Promise<U2Client> {
    if (!this.client) {
      return this.connect();
    }
    return this.client;
  }
}

File: src/u2ctl/runtime/device.py
"""Device session lifecycle and connection management."""

from typing import Optional, Any
import uiautomator2 as u2

from u2ctl.errors import DeviceOfflineError, TimeoutError
from u2ctl.runtime.adb import select_target_device, get_adb_path


class DeviceSession:
    def __init__(self, serial: Optional[str] = None, timeout: int = 30, adb_path: Optional[str] = None):
        self.serial = serial
        self.timeout = timeout
        self.adb_path = adb_path
        self._u2_device: Optional[Any] = None

    def connect(self) -> Any:
        target, _ = select_target_device(self.serial, self.adb_path)
        self.serial = target.serial
        try:
            d = u2.connect(target.serial)
            d.wait_timeout = self.timeout
            self._u2_device = d
            return d
        except Exception as first_err:
            # Single auto-reconnect retry attempt for retryable connection errors
            try:
                d = u2.connect(target.serial)
                d.wait_timeout = self.timeout
                self._u2_device = d
                return d
            except Exception as second_err:
                raise DeviceOfflineError(
                    f"Failed to connect uiautomator2 to device '{target.serial}' after retry: {second_err}"
                )

    @property
    def u2(self) -> Any:
        if self._u2_device is None:
            return self.connect()
        return self._u2_device
